scripts/safe_template_linter.py

```diff
@@ -1365,10 +1365,32 @@ def _check_concat_with_html(self, file_contents, rule, results):
         """
         lines = StringLines(file_contents)
         last_expression = None
-        # attempt to match a string that starts with '<' or ends with '>'
-        regex_string_with_html = r"""["'](?:\s*<.*|.*>\s*)["']"""
-        regex_concat_with_html = r"(\+\s*{}|{}\s*\+)".format(regex_string_with_html, regex_string_with_html)
-        for match in re.finditer(regex_concat_with_html, file_contents):
+        # Match quoted strings that starts with '<' or ends with '>'.
+        regex_string_with_html = r"""
+            {quote}                             # Opening quote.
+                (
+                   \s*<                         # Starts with '<' (ignoring spaces)
+                   ([^{quote}]|[\\]{quote})*    # followed by anything but a closing quote.
+                |                               # Or,
+                   ([^{quote}]|[\\]{quote})*    # Anything but a closing quote
+                   >\s*                         # ending with '>' (ignoring spaces)
+                )
+            {quote}                             # Closing quote.
+        """
+        # Match single or double quote.
+        regex_string_with_html = "({}|{})".format(
+            regex_string_with_html.format(quote="'"),
+            regex_string_with_html.format(quote='"'),
+        )
+        # Match quoted HTML strings next to a '+'.
+        regex_concat_with_html = re.compile(
+            r"(\+\s*{}|{}\s*\+)".format(
+                regex_string_with_html.format(quote='quote1'),
+                regex_string_with_html.format(quote='quote2')
```

Comment:
I may be wrong.  However, here's an analogy to help better describe what I mean.

Let's imagine a checkerboard where all of the black squares represent the string literals and all of the white squares are the code between the strings.  Now, let's say early on in a file we mess up what we think is a string literal.  The rest of the board will now have all of the squares flipped, where black squares are incorrectly the code between strings, and not the strings themselves, and vice versa for the white squares.

The reason why I called the more complex regex self-correcting is that if it has a false positive or false negative early on in the file, it doesn't mess with its chances of getting back on track later on in the file.

I still might be missing something, but hopefully this makes my point more clear if it wasn't clear already.